test/e2e/route_metrics_test.go

```diff
@@ -0,0 +1,393 @@
+			t.Fatalf("failed to fetch expected metrics: %v", err)
+		}
+	}
+
+	if testRS {
+		// Update the RouteSelector of the Ingress Controller so that the Route gets un-admitted again.
+		if err := kclient.Get(context.TODO(), icName, ic); err != nil {
+			t.Fatalf("failed to get ingress resource: %v", err)
+		}
+		ic.Spec.RouteSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": incorrectLabel,
+			},
+		}
+		if err := kclient.Update(context.TODO(), ic); err != nil {
+			t.Fatalf("failed to update ingresscontroller: %v", err)
+		}
+
+		// Wait for metrics to be updated to zero as the Route will get un-admitted by the IC again.
+		if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+			t.Fatalf("failed to fetch expected metrics: %v", err)
+		}
+
+		// Update the RouteSelector of the Ingress Controller so that the Route gets admitted again.
+		if err := kclient.Get(context.TODO(), icName, ic); err != nil {
+			t.Fatalf("failed to get ingress resource: %v", err)
+		}
+		ic.Spec.RouteSelector = &metav1.LabelSelector{
+			MatchLabels: map[string]string{
+				"type": correctLabel,
+			},
+		}
+		if err := kclient.Update(context.TODO(), ic); err != nil {
+			t.Fatalf("failed to update ingresscontroller: %v", err)
+		}
+
+		// Wait for metrics to be updated to 1 as the Route will get admitted by the IC again.
+		if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 1); err != nil {
+			t.Fatalf("failed to fetch expected metrics: %v", err)
+		}
+
+		// Update the label of the route so that it gets un-admitted from the Ingress Controller.
+		if err := kclient.Get(context.TODO(), routeFooLabelName, routeFooLabel); err != nil {
+			t.Fatalf("failed to get route resource: %v", err)
+		}
+		routeFooLabel.Labels = map[string]string{
+			"type": incorrectLabel,
+		}
+		if err := kclient.Update(context.TODO(), routeFooLabel); err != nil {
+			t.Fatalf("failed to update route: %v", err)
+		}
+
+		// Wait for metrics to be updated to 0 as the Route will get un-admitted by the IC.
+		if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 0); err != nil {
+			t.Fatalf("failed to fetch expected metrics: %v", err)
+		}
+
+		// Update the label of the route so that it gets admitted to the Ingress Controller again.
+		if err := kclient.Get(context.TODO(), routeFooLabelName, routeFooLabel); err != nil {
+			t.Fatalf("failed to get route resource: %v", err)
+		}
+		routeFooLabel.Labels = map[string]string{
+			"type": correctLabel,
+		}
+		if err := kclient.Update(context.TODO(), routeFooLabel); err != nil {
+			t.Fatalf("failed to update route: %v", err)
+		}
+
+		// Wait for metrics to be updated to 1 as the Route will get admitted by the IC again.
+		if err := waitForRouteMetricsAddorUpdate(t, prometheusClient, ic.Name, 1); err != nil {
+			t.Fatalf("failed to fetch expected metrics: %v", err)
+		}
+	}
+
```

Comment:
Is there a case for `testRS && testNS` ?

Reply:
But there are a different set of parameters in https://github.com/openshift/cluster-ingress-operator/pull/779/files#diff-e598c84af57afafcad05182766215625ee4aad8797344fd35381f2a244cd5211R94 for the `testRS && testNS` condition, and the code output is different for that condition too. Here you test in a way that handles that condition differently than it would be handled at runtime.

Reply:
The parameters are set to different values just to avoid the clash of same names with other tests like only `testRS` to be true or only `testNS` to be true. 

But for running the tests both the checks for `if testRS { .. }` and `if testNS { .. }`  should run for the case of `testRS && testNS`. This is the case when both `NamespaceSelector` and `RouteSelectors` are set on the IC. For the case of only `testRS` to be true only the checks for `if testRS { .. }` should be run. This is the case when only `RouteSelectors` is set on the IC. Similarly for the case of only `testNS` to be true only the checks for `if testNS { .. }` should be run. This is the case when only `NamespaceSelector` is set on the IC.